jep/0000/README.adoc

```diff
@@ -0,0 +1,232 @@
+//| :bulb: Link to where discussion and final status announcement will occur :bulb:
+//
+//
+// Uncomment if this JEP depends on one or more other JEPs.
+| Requires
+| JEP-300, JEP-303
+//
+//
+// Uncomment and fill if this JEP is rendered obsolete by a later JEP
+//| Superseded-By
+//| :bulb: JEP-NUMBER :bulb:
+//
+//
+// Uncomment when this JEP status is set to Accepted, Rejected or Withdrawn.
+//| Resolution
+//| :bulb: Link to relevant post in the jenkinsci-dev@ mailing list archives :bulb:
+
+|===
+
+
+== Abstract
+
+Integral to link:https://github.com/jenkinsci/jep/tree/master/jep/300[Jenkins
+Essentials] is the client's update lifecycle with the Evergreen hosted service
+layer. Starting for the first boot, the client must frequently inform the
+client of its current versions, retrieve new versions, and apply updates. This
+document specifies those API interactions and expectations to allow this update
+lifecycle to be performed correctly.
+
+== Specification
+
+The role of the versions service is simply to maintain an audit trail of sorts
+for the versions of sofware that each client has installed locally.
+
+
+The role of the update service is to use the `essentials.yaml` and retriev the
+latest versions of software for a given client UUID and compute the appropriate
+update response by the client
+
+.Bootstrapping Instance
+[source]
+----
+
+  update                     versions                    evergreen
+  service                    service                      client
+     |                           |                           |
+     |                           |     POST version manifest |
+     |                           |     from local hpis/core  |
+     |                           |<--------------------------o
+     |                           o-----------/ok/----------->|
+     |                           |                           |
+     |                     GET /update                       |
+     |<------------------------------------------------------o
+     |    GET /versions for uuid |                           |
+     o-------------------------->|                           |
+     |<-----/200 json payload/---o                           |
+     |                           |                           |
+     |            200 with computed JSON update manifest     |
+     |            indicating what files should               |
+     |            be downloaded                              |
+     o------------------------------------------------------>|
+     |                           |                           |
+----
+
+[[update-manifest]]
+=== Update Manifest
+
+The responses sent to the client must be well-formed JSON documents, referred
+to as "update manifests" which the client must understand. The Update Manifest
+should have a consistent structure but will be dynamically generated _per
+client_ in order to ensure that the client is only downloading what is
+necessary to update that specific client.
```

Comment:
Do you consider all updates as 'safe'?
What happened if a client didn't connect to the update service for month?
Is it an information that would be useful in the update manifest?

Reply:
This is a really interesting question! Rephrased the problem would be "what if a client was offline for a longer period of time such as not to allow for a daily upgrade, what should our behavior be then."

I'll need to think about how we should handle that some more and consult with @batmat when he's returned tomorrow (Thursday).

My current guess is that we should try to update the client "one by one" through the update manifests over time, which has some data retention requirements I hadn't previously considered.


Thought provoking question for sure @olblak  :smile_cat: